### PR TITLE
Restructure repository to solidify public interface and document it

### PR DESCRIPTION
**Pull request by coretl**
We are moving towards a 1.0 release, so should solidify our public interface. This PR documents what it should be, and then rearranges the repository accordingly.

### CODE REVIEW

File: docs/developer/explanations/decisions/0007-subpackage-structure.rst [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+from ophyd_async.epics import ADCore, ADPilatus
+
+drv = ADPilatus.ADPilatus(prefix + "DRV:")
+hdf = ADCore.NDFileHDF(prefix + "HDF:")
+```
+
+tango
+~~~~~
+
+TBC
```
Comment (callumforrester) [inline docs/developer/explanations/decisions/0007-subpackage-structure.rst:53]: Are we going to merge this as TBC?

Comment (coretl) [reply docs/developer/explanations/decisions/0007-subpackage-structure.rst:53]: Probably, I will replace "TBC" with some words that say as close to EPICS as possible...

File: docs/developer/explanations/decisions/0007-subpackage-structure.rst [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+There will be a flat public namespace under core, with contents reimported from an underscore prefixed python file. 
+Suggested something like `_device.py` containing `Device` and `DeviceVector`.
+
+epics
+~~~~~
+
+There should be an additional level of packages that correspond to the epics support module, and contain classes that map to the EPICS database and logic classes e.g.:
+
+- ``epics.signal``: containing ``epics_signal_rw``, ``EpicsTransport``, etc.
+- ``epics.ADCore``: containing ``ADDriver``, ``NDFileHDF``, ``HDFWriter`` etc.
```
Comment (coretl) [inline docs/developer/explanations/decisions/0007-subpackage-structure.rst:38]: Put in a new submodule:
`ophyd_async.epics.areadetector.adcore`

Comment (coretl) [reply docs/developer/explanations/decisions/0007-subpackage-structure.rst:38]: Stick `AD` (or `ND`) in the front of areaDetector template classes to clarify what is a bucket of PVs

**Comment by coretl** [comment]
Need to:
- decide on the strategy and merge it
- restructure the modules to ensure decent private/public namespaces
- ensure low level epics facing Devices follow the structure of templates

File: docs/explanations/decisions/0007-subpackage-structure.md [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+Detector modules should include a reference `StandardDetector` subclass, but without making any site specific decisions about PV naming.
+
+### tango
+
+This has not been created at the time of writing, but it is envisioned that it will follow the EPICS structure where it makes sense.
+
+### panda
+
+This should be a top level namespace containing the `CommonPandaBlocks`, `SeqBlock`, `PandaHdfWriter`, `PandaPcapController`, etc. These should be included in files like `_control.py`, `_writer.py`, `_block.py`, `_table.py` and imported into the top level namespace.
```
Comment (Tom-Willemsen) [inline docs/explanations/decisions/0007-subpackage-structure.md:68]: Would it be useful to briefly justify why `PandA` is special enough to be in ophyd-async directly, in contrast to most of the other devices? (I'm sure there's a good reason, I just don't know what it is, and writing the reason in this or a similar ADR would help me).

**Comment by callumforrester** [comment]
@coretl is this still up to date?

File: docs/explanations/decisions/0007-subpackage-structure.md (hunk 3 above)
Comment (coretl) [reply docs/explanations/decisions/0007-subpackage-structure.md:68]: Good question. `PandA` will eventually be one of a class of devices that are implemented in [FastCS](https://github.com/DiamondLightSource/FastCS) so therefore can use either EPICS or Tango as the transport layer.

Thinking about this, should we make a `fastcs` top level directory then? So we have `from ophyd_async.fastcs import panda` then `panda.HDFPanda`? That would leave the namespace free for other FastCS devices. Although even PandA isn't actually a FastCS device yet, so it's not technically true at the moment...

File: docs/explanations/decisions/0007-subpackage-structure.md [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+### epics
+
+Epics modules consist of 2 sorts of classes:
+- `IO` classes, which are `Device` subclasses containing `Signals` that map as closely as possible to the EPICS template hierarchy. Their name matches the EPICS template with the suffix `IO`.
+- `Control` and `Writer` classes that plug into `StandardDetector` and `StandardFlyer` telling them which logic to use during various bluesky verbs
+
+There should be an additional level of packages that correspond to the epics support module, and contain classes that map to the EPICS database and logic classes e.g.:
+
+- `epics.signal``: containing `epics_signal_rw`, `EpicsTransport`, `CaSignalBackend`, etc.
```
Comment (evalott100) [inline docs/explanations/decisions/0007-subpackage-structure.md:43]:
```suggestion
- `epics.signal`: containing `epics_signal_rw`, `EpicsTransport`, `CaSignalBackend`, etc.
```

Comment (evalott100) [reply docs/explanations/decisions/0007-subpackage-structure.md:43]: Causes weird formatting:

image

File: docs/explanations/decisions/0007-subpackage-structure.md [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+- `_detector.py` for `StandardDetector`, `DetectorWriter`, `DetectorControl`, `TriggerInfo`, etc.
+- `_flyer.py` for `StandardFlyer`, `FlyerControl`, etc.
+
+There are some renames that will be required, e.g. `HardwareTriggeredFlyable` -> `StandardFlyer`
+
+### epics
+
+Epics modules consist of 2 sorts of classes:
+- `IO` classes, which are `Device` subclasses containing `Signals` that map as closely as possible to the EPICS template hierarchy. Their name matches the EPICS template with the suffix `IO`.
+- `Control` and `Writer` classes that plug into `StandardDetector` and `StandardFlyer` telling them which logic to use during various bluesky verbs
```
Comment (evalott100) [inline docs/explanations/decisions/0007-subpackage-structure.md:39]:
```suggestion
- `Control` and `Writer` classes that plug into `StandardDetector` and `StandardFlyer` telling them which logic to use during various bluesky verbs.
```

File: docs/explanations/decisions/0007-subpackage-structure.md [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+### planstubs
+
+There will be some planstubs for shared setup that may reach across modules like `epics` and `panda`, these should live in a `planstubs/` package.
+
+### sim
+
+There will be 2 subpackages:
+- `sim.demo` for demo devices like `PatternDetector` used in tutorials
+- `sim.testing` for devices that will support tests in `ophyd-async` and `bluesky`
```
Comment (evalott100) [inline docs/explanations/decisions/0007-subpackage-structure.md:78]:
```suggestion
There will be 2 subpackages:
- `sim.demo` for demo devices like `PatternDetector` used in tutorials.
- `sim.testing` for devices that will support tests in `ophyd-async` and `bluesky`.
```

**Review by evalott100 — CHANGES_REQUESTED** [review]

**Comment by evalott100** [comment]
@coretl 
> Yes, although I would like some input from @evalott100 as to whether any of the contents of `ophyd.core._*` are actually wrong because of circular dependencies. There are bound to be some extra files, but I'd like the examples in the doc to be actually correct.

In the current state, there wouldn't be problems with circular imports from the `core._*` detailed here - `signal.py`, `sim_signal_backend.py`, and `signal_backend.py` could be aggregated to one file:

image

However my upcoming https://github.com/bluesky/ophyd-async/pull/251 wouldn't be able to adhere to this...

image

File: docs/explanations/decisions/0007-subpackage-structure.md (hunk 3 above)
Comment (callumforrester) [reply docs/explanations/decisions/0007-subpackage-structure.md:68]: Would like very careful dependency management here. A compatibility layer between two libraries should be kept independent (i.e. if you want to use it you should `pip install ophyd-async[fastcs]`)

File: docs/explanations/decisions/0007-subpackage-structure.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+This ADR proposes a public sub-package structure. The internal private structure should be flat, but can change according to the number of classes in a public package.
+
+## Decision
+
+### core
+
+There will be a flat public namespace under core, with contents reimported from an underscore prefixed python files, e.g.:
+
+- `_status.py` for `AsyncStatus`, `WatchableAsyncStatus`, etc.
+- `_protocol.py` for `AsyncReadable`, `AsyncStatus`, etc.
```
Comment (abbiemery) [inline docs/explanations/decisions/0007-subpackage-structure.md:29]: `AsyncStatus` is from `_status.py` and not `_protocol.py`.

**Review by abbiemery — CHANGES_REQUESTED** [review]

File: docs/explanations/decisions/0007-subpackage-structure.md (hunk 7 above)
Comment (abbiemery) [reply docs/explanations/decisions/0007-subpackage-structure.md:29]: Ah there is a bit of a circular thing going on here.

Comment (coretl) [reply docs/explanations/decisions/0007-subpackage-structure.md:29]:
```suggestion
- `_protocol.py` for `AsyncReadable`, `AsyncStageable`, etc.
```

File: docs/explanations/decisions/0007-subpackage-structure.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+### epics
+
+Epics modules consist of 2 sorts of classes:
+- `IO` classes, which are `Device` subclasses containing `Signals` that map as closely as possible to the EPICS template hierarchy. Their name matches the EPICS template with the suffix `IO`.
+- `Control` and `Writer` classes that plug into `StandardDetector` and `StandardFlyer` telling them which logic to use during various bluesky verbs.
+
+There should be an additional level of packages that correspond to the epics support module, and contain classes that map to the EPICS database and logic classes e.g.:
+
+- `epics.signal`: containing `epics_signal_rw`, `EpicsTransport`, `CaSignalBackend`, etc.
```
Comment (coretl) [inline docs/explanations/decisions/0007-subpackage-structure.md:47]:
```suggestion
- `epics.signal`: containing `epics_signal_rw`, `CaSignalBackend`, etc.
```

File: docs/explanations/decisions/0007-subpackage-structure.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+For PandA its namespace should contain `CommonPandaBlocks`, `SeqBlock`, `PandaHdfWriter`, `PandaPcapController`, etc. These should be included in files like `_control.py`, `_writer.py`, `_block.py`, `_table.py` and imported into the `panda` namespace.
+
+### planstubs
+
+There will be some planstubs for shared setup that may reach across modules like `epics` and `fastcs/panda`, these should live in a `planstubs/` package.
+
+### sim
+
+There will be 2 subpackages:
+- `sim.demo` for demo devices like `SimPatternDetector` and `SimMotor` used in tutorials.
```
Comment (coretl) [inline docs/explanations/decisions/0007-subpackage-structure.md:88]:
```suggestion
- `sim.demo` for demo devices like `PatternDetector` and `SimMotor` used in tutorials.
```

File: docs/explanations/decisions/0007-subpackage-structure.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+- `sim.testing` for devices that will support tests in `ophyd-async` and `bluesky`. Some test fixtures could be moved here
+
+There should probably be one file per Device, in an underscore prefixed file reimported into the public namespace, e.g. `demo/_sim_motor.py` and `demo/_sim_pattern_detector.py`.
```
Comment (coretl) [inline docs/explanations/decisions/0007-subpackage-structure.md:91]:
```suggestion
There should probably be one file per Device, in an underscore prefixed file reimported into the public namespace, e.g. `demo/_sim_motor.py` and `demo/_pattern_detector.py`.
```

**Review by abbiemery — APPROVED** [review]

**Pull request merged**